tendermint/src/block/commit_sig.rs

```diff
@@ -74,47 +77,53 @@ impl TryFrom<RawCommitSig> for CommitSig {
                 let timestamp = value.timestamp.unwrap();
                 // 0001-01-01T00:00:00.000Z translates to EPOCH-62135596800 seconds
                 if timestamp.nanos != 0 || timestamp.seconds != -62135596800 {
-                    return Err(Kind::InvalidTimestamp
-                        .context("absent commitsig has non-zero timestamp")
-                        .into());
+                    return Err(error::non_zero_timestamp_error());
                 }
             }
             if !value.signature.is_empty() {
-                return Err(Kind::InvalidSignature.into());
+                return Err(error::invalid_signature_error("invalid signature error".into()));
             }
             return Ok(CommitSig::BlockIdFlagAbsent);
         }
         if value.block_id_flag == BlockIdFlag::Commit.to_i32().unwrap() {
             if value.signature.is_empty() {
-                return Err(Kind::InvalidSignature
-                    .context("regular commitsig has no signature")
-                    .into());
+                return Err(error::invalid_signature_error("regular commitsig has no signature".into()));
             }
             if value.validator_address.is_empty() {
-                return Err(Kind::InvalidValidatorAddress.into());
+                return Err(error::invalid_validator_address_error());
             }
             return Ok(CommitSig::BlockIdFlagCommit {
                 validator_address: value.validator_address.try_into()?,
-                timestamp: value.timestamp.ok_or(Kind::NoTimestamp)?.try_into()?,
+                timestamp: value
+                    .timestamp
+                    .ok_or(error::no_timestamp_error())?
+                    .try_into()
+                    .map_err(|e | {
+                        error::in_fallible_error(e)
```

Comment:
See the documentation for `Infallible` here: https://doc.rust-lang.org/std/convert/enum.Infallible.html

Basically `Infallible` is an empty enum, which means there is no way you can ever construct a valid value for `Infallible`. So if we have a function that is given an `Infallible` value, we know that can never happen, so that part of the code will never actually get executed. As a result, we can discard an infallible value `e: Infallible` using `match e {}`.

Strictly speaking, we should never need to use `Infallible` in this case. Because a `TryFrom` with `Infallible` error is just equivalent to `Try`. So the type `Timestamp` in this case should just implement `Try` instead of `TryFrom`:


```diff
--- a/tendermint/src/time.rs
+++ b/tendermint/src/time.rs
- impl TryFrom<Timestamp> for Time {
-     type Error = Infallible;
- 
-     fn try_from(value: Timestamp) -> Result<Self, Self::Error> {
-         // prost_types::Timestamp has a SystemTime converter but
-         // tendermint_proto::Timestamp can be JSON-encoded
-         let prost_value = prost_types::Timestamp {
-             seconds: value.seconds,
-             nanos: value.nanos,
-         };
- 
-         Ok(SystemTime::from(prost_value).into())
-     }
- }
+ impl From<Timestamp> for Time {
+     fn from(value: Timestamp) -> Self {
+         // prost_types::Timestamp has a SystemTime converter but
+         // tendermint_proto::Timestamp can be JSON-encoded
+         let prost_value = prost_types::Timestamp {
+             seconds: value.seconds,
+             nanos: value.nanos,
+         };
+ 
+         SystemTime::from(prost_value).into()
+     }
+ }
```

Then here we can just rewrite it as

```rust
timestamp: value
    .timestamp
    .ok_or(error::no_timestamp_error())?
    .into()
```

See the commit https://github.com/informalsystems/ibc-rs/pull/988/commits/6b397f183c7ceb1d3bbf6f396b4f208590dbbf11 for similar work done at https://github.com/informalsystems/ibc-rs/pull/988.

Reply:
I understand thank you.